source/meshworker/scratch_data.cc

```diff
@@ -0,0 +1,363 @@
+    , fe(&fe)
+    , cell_quad(quadrature)
+    , face_quad(face_quadrature)
+    , cell_flags(update_flags)
+    , neighbor_cell_flags(neighbor_update_flags)
+    , face_flags(face_update_flags)
+    , neighbor_face_flags(neighbor_face_update_flags)
+    , local_dof_indices(fe.dofs_per_cell)
+    , neighbor_dof_indices(fe.dofs_per_cell)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  ScratchData<dim, spacedim>::ScratchData(
+    const FiniteElement<dim, spacedim> &fe,
+    const Quadrature<dim> &             quadrature,
+    const UpdateFlags &                 update_flags,
+    const Quadrature<dim - 1> &         face_quadrature,
+    const UpdateFlags &                 face_update_flags)
+    : ScratchData(StaticMappingQ1<dim, spacedim>::mapping,
+                  fe,
+                  quadrature,
+                  update_flags,
+                  face_quadrature,
+                  face_update_flags)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  ScratchData<dim, spacedim>::ScratchData(
+    const FiniteElement<dim, spacedim> &fe,
+    const Quadrature<dim> &             quadrature,
+    const UpdateFlags &                 update_flags,
+    const UpdateFlags &                 neighbor_update_flags,
+    const Quadrature<dim - 1> &         face_quadrature,
+    const UpdateFlags &                 face_update_flags,
+    const UpdateFlags &                 neighbor_face_update_flags)
+    : ScratchData(StaticMappingQ1<dim, spacedim>::mapping,
+                  fe,
+                  quadrature,
+                  update_flags,
+                  neighbor_update_flags,
+                  face_quadrature,
+                  face_update_flags,
+                  neighbor_face_update_flags)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  ScratchData<dim, spacedim>::ScratchData(
+    const ScratchData<dim, spacedim> &scratch)
+    : mapping(scratch.mapping)
+    , fe(scratch.fe)
+    , cell_quad(scratch.cell_quad)
+    , face_quad(scratch.face_quad)
+    , cell_flags(scratch.cell_flags)
+    , neighbor_cell_flags(scratch.neighbor_cell_flags)
+    , face_flags(scratch.face_flags)
+    , neighbor_face_flags(scratch.neighbor_face_flags)
+    , local_dof_indices(scratch.local_dof_indices)
+    , neighbor_dof_indices(scratch.neighbor_dof_indices)
+    , any_data(scratch.any_data)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  const FEValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell)
+  {
+    if (!fe_values)
+      fe_values = std::make_unique<FEValues<dim, spacedim>>(*mapping,
+                                                            *fe,
+                                                            cell_quad,
+                                                            cell_flags);
+
+    fe_values->reinit(cell);
+    cell->get_dof_indices(local_dof_indices);
+    add_ref<FEValuesBase<dim, spacedim>>(*fe_values, "FEValuesBase");
```

Comment:
In deal.II, all neighbors are with no u. I just did what was already there... :)

Reply:
Would moving the boost::any data related stuff into `dealii::AnyData` class, and instantiating one for internal use, and one for external use work for you?

Users would call `cache.get_object_with_name()`, and internally this would be forwarded to a `user_cache` object of type `AnyData`, while all the rest would be stored in a `internal_cache` object of the same type.

Reply:
I didn't read the code in detail, but these functions seem to suggest that you are reinventing something like `boost::any`.

`AnyData` is implementing similar functionality, and internally uses `boost::any`.

Reply:
> In deal.II, all neighbors are with no u. I just did what was already there... :)

I know, but I was trying to point out that its easy to introduce a spelling error (if extending the class at a later stage) into the hard-coded names :-D

Anyway, since you've moved these to their own smart pointers its a moot point :-)

Reply:
> Would moving the boost::any data related stuff into dealii::AnyData class, and instantiating one for internal use, and one for external use work for you?

> Users would call cache.get_object_with_name(), and internally this would be forwarded to a user_cache object of type AnyData, while all the rest would be stored in a internal_cache object of the same type.

Sorry, I didn't address these comments. Yes, I think that this might be the best way forward. That would ensure that any (unlikely) name clashes would be completely avoided.

I must admit that I wasn't aware of the `AnyData` class until now. Given the (long standing) TODO sitting in its documentation, I wonder how much its being used anyway... Yet, its a really useful class that we could for sure make more regular use of (e.g. `QuadratureHistory`). 

I propose that in another PR we finally update it using what @luca-heltai has integrated into the `ScratchData` class.  Although at a glance the implementations look quite similar, since @luca-heltai has already got the map-based implementation sitting here we could really just move this over with little modification.
In summary, I propose that we: 
- update its interface with that which has been developed here (`add_*`, `get_*`, `stores_*`)
- add `try_get_*` functions that would be the equivalent of the `try_read_*` functions, and
- finally deprecate the index-based access, and switch it over to a map-based data structure.

If this is the agreed way forward, then @luca-heltai I'd say that we can hold off on the split that adds an internal and user-only data structure until then.